src/main/java/com/transferwise/mitosis/SeoExperiment.java

```diff
@@ -0,0 +1,22 @@
+package com.transferwise.mitosis;
+
+import javax.servlet.http.HttpServletRequest;
+import java.util.List;
+import java.util.function.Predicate;
+
+public class SeoExperiment extends Experiment {
+
+    public SeoExperiment(String name, List<String> variants, Predicate<HttpServletRequest> filter) {
+        super(name, variants, filter);
+    }
+
+    public SeoExperiment(String name, List<String> variants) {
+        super(name, variants, null);
+    }
+
+    @Override
+    public String chooseVariant(HttpServletRequest request) {
+        int index = Math.abs(request.getServletPath().hashCode()) % variants.size();
```

Comment:
It seems we'll still be fairly uniform if we set a threshold and use that instead of `variants.size()`:
`(Math.abs(request.getServletPath().hashcode()) % 10) <= threshold`

This would allow for a risk-averse test—for example if the threshold was 10 we could apply the change to only 10% of pages.

But that would mean the order of the variants would matter (say: first is always the control, and second the variant), and `threshold` would be an additional argument, and it still doesn't scale to multiple buckets of URLs if we wanted to do that. For example, three buckets of equal size.

If the user/client specifies they want:
- bucket1 with 33% of URLs
- bucket2 with 33% of URLs
- bucket3 with 33% of URLs

is there an efficient way to map the result of `(Math.abs(request.getServletPath().hashcode()) % something)` to one of those buckets?

If so, Mitosis clients could pass a Hashmap representing their variants and bucket sizes, rather than a `List<String>`, and so you'll still have just three parameters: `name`, `variants`, `filter`, which is a nice signature.

Reply:
I don't understand the distinction between variants and buckets in your example.

If the user/client specifies three variants then (given a large enough number of URLs) 
33% of URLs will be assigned to variant 1
33% of URLs will be assigned to variant 2
33% of URLs will be assigned to variant 3

Reply:
As for the threshold I think you can achieve that using filters. The following test will print out the number of tests out of 1000 the Experiment was applied to. And in this case it should be roughly 100. Because the Experiment contains only only one variant every url that passes the threshold will be assigned the same variant.

```
    def 'threshold' () {
        given:
        Map<String, String> variants = new HashMap<String, String>();
        int count = 0;

        engine.register(new SeoExperiment(
                'test',
                ['a'],
                new Predicate<HttpServletRequest>() {
                    @Override
                    boolean test(HttpServletRequest request) {
                        int threshold = Math.abs(request.getServletPath().hashCode()) % 100
                        return threshold < 10
                    }
                })
        );

        when:

        for (int i = 0; i < 1000; i++) {
            variants = engine.refreshVariants(variants, createRequestMock("path" + i))
            count += variants.size();
        }

        then:
        println count

    }
```

Reply:
Here's a test that shows the 10% of urls that pass the threshold get the variants evenly split among them:

```
    def 'threshold' () {
        given:
        Map<String, String> variants = new HashMap<String, String>();
        int aCount = 0;
        int bCount = 0;

        engine.register(new SeoExperiment(
                'test',
                ['a', 'b'],
                new Predicate<HttpServletRequest>() {
                    @Override
                    boolean test(HttpServletRequest request) {
                        int threshold = Math.abs(request.getServletPath().hashCode()) % 100
                        return threshold < 10
                    }
                })
        );

        when:

        for (int i = 0; i < 1000; i++) {
            variants = engine.refreshVariants(variants, createRequestMock("path" + i))
            if (variants.size() > 0) {
                if (variants.get('test') == 'a') {
                    aCount++
                } else {
                    bCount++
                }
            }
        }

        then:
        println aCount
        println bCount

    }
```